### PR TITLE
Added a feature that replaces tikzpicture/pgfplot source code by .pdf figures

### PR DESCRIPTION
**Pull request by sdnr**
Hi,

This is my first ever pull request on github, so please be patient with me :-)

I was always angry at arxiv when some people used my/our tikz source code (that we spent several work hours on) within their papers and simply added an additional node or whatever, without even citing our paper for the figure.
Yes I know, I also like the idea of open science and all of that, but I don't like it when people copy my "design/style", which I feel is like copying whole sentences of my paper.
Also sometimes the meta data of plots can reveal who conducted the simulation/measurement by e.g. its file name.
Or sometimes other information that you don't want to share with others remains in raw data used within pgfplots.

This is why I always replaced all tikzpicture source code within my papers by hand with a simple `\includegraphics` of the figure's pdf that was created automatically when I used tikz' externalize package `\usetikzlibrary{external}`.
As I am using your arxiv-latex-cleaner (kudos for that, great tool) I thought this would be a nice feature to add.

So the requirements are, that all the figures you want to autoreplace, are precompiled as .pdf and lie in a folder within the main folder.
For example:
- one needs to add `\tikzsetnextfilename{picture_name}` in front of every `\begin{tikzpicture}` environment first. This is to later be able to identify the figure, because figures could be defined in several included tex files (without order).
- then one needs to compile the tex code with tikz external package, by simply adding `\usetikzlibrary{external}` to the tex preamble and with `\tikzexternalize[prefix=ext_tikz/]` one can define a folder where all externally compiled figures are stored. See section 53 in the PGF/TikZ manual (https://ctan.org/pkg/pgf?lang=en) on TikZ picture externalization.
- if these requirements are made, one can simply use the feature by adding the argument `--use_external_tikz ext_tikz`

With this, thanks to main functions of arxiv-latex-cleaner, only the included pdf files will remain in the arxiv output and all included raw data files and tikz source code are kicked out.

I also added an update to the README and added some unit tests, but I am not a professional programmer so you should maybe have a look at them.

I hope you like this feature. Just tell me if you want me to change something.

Best,
Sebastian

### CODE REVIEW

**Comment by sdnr** [comment]
@googlebot I signed it!

File: README.md [hunk 1]
```diff
@@ -92,6 +105,9 @@ optional arguments:
   --commands_to_delete COMMANDS_TO_DELETE [COMMANDS_TO_DELETE ...]
                         LaTeX commands that will be deleted. Useful for e.g.
                         user-defined \todo commands.
+  --use_external_tikz EXTERNAL_TIKZ_FOLDER
+                        Folder (relative to input folder) containing
+                        externalized TikZ figures in PDF format
```
Comment (jponttuset) [inline README.md:110]:
```suggestion
                        externalized TikZ figures in PDF format.
```

File: README.md [hunk 2]
```diff
@@ -57,6 +57,19 @@ There is a 10MB limit on arXiv submissions, so to make it fit:
     You can whitelist some PDFs to skip the global size using
     `images_whitelist`.
 
+#### TikZ picture source code concealment
+
+To prevent the upload of carefully crafted tikzpicture source code or raw simulation
```
Comment (jponttuset) [inline README.md:62]:
```suggestion
To prevent the upload of tikzpicture source code or raw simulation
```
I'd remove 'carefully crafted' as the tone implies that one does not want to share something just because it took time to create :)
I know it's part of your spirit, but let's leave it more vague here, as there might be many other reasons people cannot share their tikz code.

File: arxiv_latex_cleaner/__main__.py [hunk 3]
```diff
@@ -83,6 +83,10 @@
           "\\todo commands."),
 )
 
+PARSER.add_argument(
+    "--use_external_tikz", type=str, help="Folder (relative to input folder) containing externalized tikz figures in PDF format"
```
Comment (jponttuset) [inline arxiv_latex_cleaner/__main__.py:87]:
```suggestion
    "--use_external_tikz", type=str, help="Folder (relative to input folder) containing externalized tikz figures in PDF format."
```

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py [hunk 4]
```diff
@@ -151,11 +151,31 @@ def _remove_comments(content, parameters):
...
   for command in parameters['commands_to_delete']:
     content = _remove_command(content, command)
-  # If file ends with '\n' already, the split in last line would add an extra
-  # '\n', so we remove it.
-  return content.split('\n')
+  return content
 
 
+def _replace_tikzpictures(content, figures):
+  """Replaces all tikzpicture environments (with includegraphic commands of external PDF figures) in the content, and writes it."""
```
Comment (jponttuset) [inline arxiv_latex_cleaner/arxiv_latex_cleaner.py:158]: Please keep the lines to 80 characters long.

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py [hunk 5]
```diff
@@ -151,11 +151,31 @@ def _remove_comments(content, parameters):
...
+
+  def get_figure(matchobj):
+#      print("tikzpicture: match found ", matchobj.group(0))
```
Comment (jponttuset) [inline arxiv_latex_cleaner/arxiv_latex_cleaner.py:161]:
```suggestion
```

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py [hunk 6]
```diff
@@ -151,11 +151,31 @@ def _remove_comments(content, parameters):
...
+      found_tikz_filename = re.search(r'\\tikzsetnextfilename{(.*?)}',matchobj.group(0)).group(1)
```
Comment (jponttuset) [inline arxiv_latex_cleaner/arxiv_latex_cleaner.py:162]:
```suggestion
      found_tikz_filename = re.search(r'\\tikzsetnextfilename{(.*?)}', matchobj.group(0)).group(1)
```

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py [hunk 7]
```diff
@@ -151,11 +151,31 @@ def _remove_comments(content, parameters):
...
+#      print("tikzpicture: looking for filename ", found_tikz_filename)
```
Comment (jponttuset) [inline arxiv_latex_cleaner/arxiv_latex_cleaner.py:163]: Please erase all dead/debug code.

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py [hunk 8]
```diff
@@ -151,11 +151,31 @@ def _remove_comments(content, parameters):
   content = _remove_environment(''.join(content), 'comment')
   for command in parameters['commands_to_delete']:
     content = _remove_command(content, command)
-  # If file ends with '\n' already, the split in last line would add an extra
-  # '\n', so we remove it.
-  return content.split('\n')
+  return content
```
Comment (jponttuset) [inline arxiv_latex_cleaner/arxiv_latex_cleaner.py:154]: Why did you move this?

File: README.md (hunk 2 above)
Comment (sdnr) [reply README.md:62]: sure ;-)

File: arxiv_latex_cleaner/arxiv_latex_cleaner.py (hunk 8 above)
Comment (sdnr) [reply arxiv_latex_cleaner/arxiv_latex_cleaner.py:154]: I need a joined string in `_replace_tikzpictures` anyway so i performed the split after `_replace_tikzpictures`.
Otherwise I would need to add "\n" to all lines in content again, then joint them again in `_replace_tikzpictures`, and then split them again.
So I moved the content splitting part out of both functions, but yeah it's not that pretty that `_remove_comments` now consumes a content array of lines but returns a joined string.

**Comment by sdnr** [comment]
Hi @jponttuset 
I added your suggestions using github's interface and also pushed a commit that removes the debug code.
I guess the googlebot CLA issue is caused by you beeing co-author of the first commit, isn't it?

**Comment by jponttuset** [comment]
Hi @sdnr 
Thanks for your changes!
Yes, the CLA is complaining about me now, I'll fix it ASAP and merge the PR.

**Comment by sdnr** [comment]
You're welcome, Thank you for this handy tool :-)

**Comment by jponttuset** [comment]
@googlebot I consent.
